libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -0,0 +1,478 @@
+        [ r
+          | SJust r <-
+              [ requireTimeStart <$> mBefore,
+                requireTimeExpire <$> mAfter,
+                SJust requireSignature
+              ]
+        ]
+      requireSignature = RequireSignature <$> genKeyHash
+      requireAllOf k = do
+        n <- lift nonNeg1Digit
+        RequireAllOf . Seq.fromList <$> replicateM n (genNestedTimelock (k - 1))
+      requireAnyOf k = do
+        n <- lift pos1Digit
+        RequireAnyOf . Seq.fromList <$> replicateM n (genNestedTimelock (k - 1))
+      requireMOf k = do
+        n <- lift nonNeg1Digit
+        m <- lift $ choose (0, n)
+        RequireMOf m . Seq.fromList <$> replicateM n (genNestedTimelock (k - 1))
+      requireTimeStart (SlotNo validFrom) = do
+        minSlotNo <- lift $ choose (minBound, validFrom)
+        pure $ RequireTimeStart (SlotNo minSlotNo)
+      requireTimeExpire (SlotNo validTill) = do
+        maxSlotNo <- lift $ choose (validTill, maxBound)
+        pure $ RequireTimeExpire (SlotNo maxSlotNo)
+  tlscript <- genNestedTimelock (2 :: Natural)
+  let corescript :: Core.Script era
+      corescript = case proof of
+        Babbage _ -> TimelockScript tlscript
+        Alonzo _ -> TimelockScript tlscript
+        Mary _ -> tlscript
+        Allegra _ -> tlscript
+        Shelley _ -> error "Shelley does not have TimeLock scripts"
+  let scriptHash = hashScript @era corescript
+  modify $ \ts@GenState {gsScripts} -> ts {gsScripts = Map.insert scriptHash corescript gsScripts}
+  pure scriptHash
+
+-- Adds to gsScripts
+genMultiSigScript :: forall era. Reflect era => Proof era -> GenRS era (ScriptHash (Crypto era))
+genMultiSigScript proof = do
+  let genNestedMultiSig k
+        | k > 0 =
+            elementsT $
+              nonRecTimelocks ++ [requireAllOf k, requireAnyOf k, requireMOf k]
+        | otherwise = elementsT nonRecTimelocks
+      nonRecTimelocks = [requireSignature]
+      requireSignature = Shelley.RequireSignature <$> genKeyHash
+      requireAllOf k = do
+        n <- lift nonNeg1Digit
+        Shelley.RequireAllOf <$> replicateM n (genNestedMultiSig (k - 1))
+      requireAnyOf k = do
+        n <- lift pos1Digit
+        Shelley.RequireAnyOf <$> replicateM n (genNestedMultiSig (k - 1))
+      requireMOf k = do
+        n <- lift nonNeg1Digit
+        m <- lift $ choose (0, n)
+        Shelley.RequireMOf m <$> replicateM n (genNestedMultiSig (k - 1))
+  msscript <- genNestedMultiSig (2 :: Natural)
+  let corescript :: Core.Script era
+      corescript = case proof of
+        Shelley _ -> msscript
+        _ -> error (show proof ++ " does not have MultiSig scripts")
+  let scriptHash = hashScript @era corescript
+  modify $ \ts@GenState {gsScripts} -> ts {gsScripts = Map.insert scriptHash corescript gsScripts}
+  pure scriptHash
+
+primaryLanguages :: Proof era -> [Language]
+primaryLanguages (Babbage _) = [PlutusV2]
+primaryLanguages (Alonzo _) = [PlutusV1]
+primaryLanguages _ = [PlutusV1]
```

Comment:
Shouldn't this be empty, since Plutus was introduced with Alonzo?
```suggestion
primaryLanguages _ = []
```

Reply:
I think I am going to restructure this.

Reply:
We only use this in one place, and it depends on there being a Language for every era. It is just ignored for eras before Alonzo. I am going to redo it with a maybe type.